klippy/extras/ads1x1x.py

```diff
@@ -0,0 +1,431 @@
+                'samples_per_second', 128)
+
+        self._ppins = self._printer.lookup_object("pins")
+        self._ppins.register_chip(self.name, self)
+
+        self.pga = config.getchoice('pga', ADS1X1X_PGA, 'V4.096')
+        self.mode = config.getchoice('mode', ADS1X1X_MODE, 'single')
+        # Comparators are not implemented, they would only be useful if the
+        # alert pin is used, which we haven't made configurable.
+        # But that wouldn't be useful for a normal temperature sensor anyway.
+        self.comp_mode = ADS1X1X_COMPARATOR_MODE['TRADITIONAL']
+        self.comp_polarity = ADS1X1X_COMPARATOR_POLARITY['ACTIVE_LO']
+        self.comp_latching = ADS1X1X_COMPARATOR_LATCHING['NON_LATCHING']
+        self.comp_queue = ADS1X1X_COMPARATOR_QUEUE['QUEUE_NONE']
+        self._i2c = bus.MCU_I2C_from_config(config, address)
+
+        self.adc_voltage = config.getfloat('adc_voltage', 5., above=0.)
+
+        self.mcu = self._i2c.get_mcu()
+
+        self._printer.add_object("ads1x1x " + self.name, self)
+        self._printer.register_event_handler("klippy:ready", \
+                                            self._handle_ready)
+        self._printer.register_event_handler("klippy:shutdown", \
+                                            self.reset_all_devices)
+
+        self._pins = {}
+        self._command_queue = []
+
+    def setup_pin(self, pin_type, pin_params):
+        if pin_type == 'adc':
+            pin = ADS1X1X_pin(self, pin_params)
+            if pin.pin in self._pins:
+                raise pins.error(
+                    '%s pin %s for chip %s is used multiple times' \
+                        % (self.chip, pin.pin, self.name))
+            self._pins[pin.pin] = pin
+            return pin
+        raise pins.error('Wrong pin or incompatible type: %s with type %s! ' % (
+            pin_params['pin'], pin_type))
+
+    def _handle_ready(self):
+        self.reset_all_devices()
+        self.sample_timer = \
+                self._reactor.register_timer(self._sample)
+        self._reactor.update_timer(self.sample_timer, self._reactor.NOW)
+
+    def request_sample(self, sensor, callback):
+        if len(self._command_queue) <= ADS1X1_MAX_COMMAND_QUEUE_SIZE:
+            self._command_queue.append({
+                'sensor': sensor,
+                'callback': callback
+            })
+        else:
+            logging.warning(
+                'ADS1X1X: command queue is overflowing, dropping requests')
+
+    def _sample(self, eventtime):
+        if self._command_queue:
+            command = self._command_queue.pop()
+            pin_object = self._pins[command['sensor'].pin]
+            sample = 0
+            try:
+                self._write_register(ADS1X1X_REG_POINTER['CONFIG'],
+                                     pin_object.config)
+
+                self._reactor.pause(self._reactor.monotonic() + 1.0 /
+                                    (self.samples_per_second_numeric + 1))
```

Comment:
Am I right in thinking this wouldn't pause for a long enough time for the ADC conversion to complete, based on `1.0 / (self.samples_per_second_numeric + 1)`?

For example, if we were taking 128 samples a second, then the ADC delay would be 7.8125ms per sample, but `1.0 / (128 + 1)` would be roughly 7.75ms.

Would it make sense to change this to `1.0 / self.samples_per_second_numeric + 0.001`?

Reply:
According to the spec, bit 15 is a operational status pin which we can check to see whether the value's available.

> For a read status:
> 0 : Device is currently performing a conversion
> 1 : Device is not currently performing a conversion

Perhaps it's worth treating the delay as a timeout instead, and polling the config to see whether bit 15 is 1?

Reply:
I've tested this and it works very well, solving the first issue I describe in https://github.com/Klipper3d/klipper/pull/6584#issuecomment-2240574284.

https://github.com/jackwakefield/klipper/commit/72e923761c65a357adadb823f5f40df697abd8b8#diff-a00870c95aa335ed2e7afd38885cc50c69987c1d819811014a80c380f85bdd95R246

Reply:
The intent was to have a little extra delay, that's why added the +1 on the samples, but it should've been a -1 actually. I saw you are using a fixed delay now, the values seem like they are related to the default sample count for that device, not the configured one.
ADS1115 default is 128, 1 / 128 ~= 0.008
ADS1015 default is 1600, 1 / 1600 ~= 0.001
Theoretically, the sample count should relate to how long the operation takes, I'm gonna test it out, but I think instead of the fixed delay we'll go back to 1 / sampleCount but with a better margin.

Checking the conversion bit seems like a fine adjustment, an extra layer of security here is nice.

I'll have to check some stuff, but I would then cherry-pick your commit over.

Reply:
I've now updated the code, seems to be working fine. On my raspberry pi 3b, the sample rate seems to cap out somwhere at around 80 fps, but lower values seem to be working fine, too.